Make participant encoder/decoder access thread-safe

woutd opened the pull request:
Replaces gbooleans used for encoding/decoding status with mutexes to ensure all access to the participant encoder and decoder is thread-safe. Additionally, the reset status gboolean is now accessed atomically.

Currently only tested with participants using Opus encoder/decoder.

This needs to be carefully reviewed for performance issues and logic errors.

See also #3587

Commit:
Fix: Make participant encoder/decoder access thread-safe

Replaces gbooleans used for encoding/decoding status with mutexes to ensure all access to the participant encoder and decoder is thread-safe.
Additionally, the reset status gboolean is now accessed atomically.

januscla commented:
Thanks for your contribution, @woutd! Please make sure you sign our [CLA](https://cla.conf.meetecho.com/janus), as it's a required step before we can merge this.

lminiero commented on src/plugins/janus_audiobridge.c:
```diff
@@ -4900,7 +4902,7 @@ static json_t *janus_audiobridge_process_synchronous_request(janus_audiobridge_s
 			goto prepare_response;
 		}
-		participant->reset = TRUE;
+		g_atomic_int_set(&participant->reset, TRUE);
```
For atomic ints, we use numbers and not boolean values, so better have `1` there. No actual change in functionality, just code style.

lminiero commented on src/plugins/janus_audiobridge.c:
```diff
@@ -6326,32 +6328,32 @@ void janus_audiobridge_incoming_rtp(janus_plugin_session *handle, janus_plugin_r
 	janus_audiobridge_participant *participant = (janus_audiobridge_participant *)session->participant;
-	if(!g_atomic_int_get(&participant->active) || participant->muted || g_atomic_int_get(&participant->suspended) ||
-			(participant->codec == JANUS_AUDIOCODEC_OPUS && !participant->decoder) || !participant->room)
+	if(!g_atomic_int_get(&participant->active) || participant->muted || g_atomic_int_get(&participant->suspended) || !participant->room)
```
Why did you remove the check on the decoder there?

woutd replied:
At first, I added the following code above it:
```
gboolean opus_decoding = FALSE;
janus_mutex_lock(&participant->decoding_mutex);
if(participant->decoder)
	opus_decoding = TRUE;
janus_mutex_unlock(&participant->decoding_mutex);
```
And then used `opus_decoding` instead of  `participant->decoder` in the checks. But this would require an extra lock each time.

I also though of just keeping it like it was, but that could cause some memory fence issue (thread uses stale value) I think?

Since the packet would be dropped later on if `participant->decoder` is NULL, I removed the extra check.

lminiero reviewed (commented):
I've done a first quick review, so you can find some notes and comments inline, thanks!

Commit:
Code styling

Commit:
Prevent dropping all packets when we are decoding G.711

Commit:
Do not lock when we are encoding G.711

lminiero reviewed (commented):
Apologies if this took a while, I checked your changes and, apart from the considerations we made on whether it makes more sense to keep two distinct mutexes for encoding/encoding, or use just one, it looks fine to me. I haven't made any test yet, but I think it's ok to remove the draft status for now, and let people play with it to provide feedback. Did you keep on testing this in your own setup, by the way? Anything worth noting or has it worked as expected so far?

Commit:
Unlock decoding_mutex before calling janus_audiobridge_participant_istalking in janus_audiobridge_participant_thread.
This can result in a deadlock because participant->decoding_mutex and audiobridge->mutex are locked in reverse order in janus_audiobridge_hangup_media_internal.

lminiero commented on src/plugins/janus_audiobridge.c:
```diff
@@ -9152,8 +9167,10 @@ static void *janus_audiobridge_participant_thread(void *data) {
 					if(!first && participant->codec == JANUS_AUDIOCODEC_OPUS && lost_packets_gap <= JITTER_BUFFER_MAX_GAP_SIZE && !participant->muted) {
 						lost_packets_gap++;
-						if(!g_atomic_int_compare_and_exchange(&participant->decoding, 0, 1)) {
+						janus_mutex_lock(&participant->decoding_mutex);
```
No, I don't like this approach: plenty of things can happen between when you release and lock again. A better approach is to avoid the lock in that inner function if we have the mutex already (something we do in a few other places). It may also be we can get rid of that lock in the function altogether, if this is the only place we call it from. I'm travelling today so I'll check tomorrow.

woutd replied:
I do not think `participant->decoding_mutex` is locked at that point, only after.

lminiero replied:
Argh, you're right again, I was looking at the main codebase to see where the istalking function was, rather than your patch... Three days at the Faroe Islands and I've forgotten how to check code :grin:

woutd commented:
Thanks for the info @spscream

How many users were active during this peak?
Without the extra locks for thread safety you can handle the same amount of users without a noticeable extra CPU load?

I can understand the load is higher with the locks, but would expect most of the load comes from the OPUS encoding itself.

@lminiero Are there tools available to simulate load/users on the audio bridge?

spscream commented:
> How many users were active during this peak?

~150 ab publishers during peaks. Cpu usage doesn't grow evenly with participants count. Rather, it's sporadic, occurring at random points in time, and doesn't correlate with the number of participants.

woutd commented:
I simulated over 1000 audio bridge participants and cannot see any noticeable difference in CPU usage with or without the multi thread fixes.

I can also see the "Participant queue-in contains too many packets, clearing now (count=5)" log when reaching the server limit. But I also see this log without the fixes around the same number of participants.

As far as I can tell, Janus seems to have more trouble handling all the WebSocket connections than it has issues with handling the AudioBridge participants.

lminiero commented:
@woutd could you align the code to the latest changes we made in master? I plan to make a few load tests and in case all looks good merge, but considering the code the PR is based on may be older I want to make sure it includes all the latest changes (including the security fixes I merged yesterday).

woutd committed:
Merge branch 'meetecho:master' into audiobridge-multithread-fixes

woutd commented:
I have merged it with master. Please let me know if you rather have it rebased and squashed to 1 commit.

lminiero commented:
> I have merged it with master. Please let me know if you rather have it rebased and squashed to 1 commit.

Thanks for doing it so quickly! No need to rebase, as I'll squash when merging anyway (it's what I usually do, especially for branches with many commits).

lminiero commented:
> I can also see the "Participant queue-in contains too many packets, clearing now (count=5)" log when reaching the server limit. But I also see this log without the fixes around the same number of participants.

From a couple of tests I made configuring master and this branch the same way, I can confirm this is what I see too. That log lone is indeed indication that the server is struggling to keep up, but it started appearing at about the same number of participants in both branches. I didn't make a ton of tests, but this seems to align to your experience.

@spscream is there any other data you can share?

spscream commented:
@lminiero we don't have any problems with it on our production loads. Only one customer had issue with it, but no additional data or tests we can provide. I think you can merge it safely.

lminiero commented:
@spscream thanks, that helps! Merging then :muscle: 
Thank you both for the help, and the patience!
